### ISSUE TITLE
Add support for custom compressor in tarfile

### ISSUE THREAD

**Issue by 81a998e9-5ed4-4e57-9a86-eddf1d585580** [python/cpython#75203]
BPO | [31020](https://bugs.python.org/issue31020)
--- | :---
Nosy | @gustaebel, @serhiy-storchaka, @Mariatta, @insomniacslk
PRs | <li>python/cpython#2734</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2017-07-24.21:05:21.501>
labels = ['3.7', 'type-feature', 'library']
title = 'Add support for custom compressor in tarfile'
updated_at = <Date 2017-07-28.08:25:38.113>
user = 'https://github.com/insomniacslk'
```

bugs.python.org fields:
```python
activity = <Date 2017-07-28.08:25:38.113>
actor = 'insomniacslk'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['Library (Lib)']
creation = <Date 2017-07-24.21:05:21.501>
creator = 'insomniacslk'
dependencies = []
files = []
hgrepos = []
issue_num = 31020
keywords = []
message_count = 5.0
messages = ['299017', '299356', '299357', '299360', '299375']
nosy_count = 4.0
nosy_names = ['lars.gustaebel', 'serhiy.storchaka', 'Mariatta', 'insomniacslk']
pr_nums = ['2734']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'enhancement'
url = 'https://bugs.python.org/issue31020'
versions = ['Python 3.7']
```

</p></details>

**Comment by 81a998e9-5ed4-4e57-9a86-eddf1d585580** [comment]
Tarfile would benefit from exposing custom compressors. At the moment the only way to use something that is not gzip/bzip/lzma is to separate the archiving and compression steps.

A possible approach is to pass a custom compression function to `tarfile.TarFile.open`. However the current interface is not clean enough to be exposed. I have made a very conservative change via a pull request on GitHub, see https://github.com/python/cpython/pull/2734 . Some additional considerations can be found there.

A further step could require a simplified interface that only involves file name, file-like object and compression level, and returns a file-like object to read the compressed data from. For example:

```py
def my_compressor(name, fileobj=None, compresslevel=9):
    # compression happens here
    return filelike_object
```



This further step is not captured in the pull request, but I can iterate and update the diff.

**Comment by Mariatta** [comment]
Adding lars.gustaebel who is listed as maintainer for tarfile.

**Comment by Mariatta** [comment]
Adding serhiy.storchaka who recently made changes to tarfile.

**Comment by serhiy-storchaka** [comment]
You can pass a custom file object that implements a custom compression to the TarFile constructor. I don't see a need in other way.

**Comment by 81a998e9-5ed4-4e57-9a86-eddf1d585580** [comment]
The documentation should be updated to suggest that a custom file-like object can be passed to have custom compressors then.

However this sounds very hacky and definitely not a feature that can be easily passed to the user. Hence the reason for a cleaner, explicitly advertised interface.

**Comment by encukou** [comment]
I agree with Serhiy: passing a [ZipFile](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile), [BZ2File](https://docs.python.org/3/library/bz2.html#bz2.BZ2File), [LZMAFile](https://docs.python.org/3/library/lzma.html#lzma.LZMAFile), etc. composes more nicely than making `tarfile` aware of custom compressors. We need to keep the current interface for backwards compatibility, but we don't need to extend it.

Given that there's currently no dedicated maintainer for `tarfile`, adding this feature doesn't sound like a good idea.

Sorry for the delay in responding. (There is, after all, no dedicated `tarfile` maintainer.)

**Comment by picnixz** [comment]
I will close this issue as `not planned` for now and would advise creating a post on https://discuss.python.org/c/ideas/6 to gather support from the community. We'll re-open the issue once this feature is deemed relevant.
